subsys/ipc/ipc_service/lib/icmsg.c

```diff
@@ -70,54 +73,72 @@ static bool is_endpoint_ready(struct icmsg_data_t *dev_data)
 	return atomic_get(&dev_data->state) == ICMSG_STATE_READY;
 }
 
-static bool is_send_buffer_reserved(struct icmsg_data_t *dev_data)
+static bool is_tx_buffer_reserved(struct icmsg_data_t *dev_data)
 {
-	return atomic_get(&dev_data->send_buffer_reserved) ==
-			SEND_BUFFER_RESERVED;
+	return atomic_get(&dev_data->tx_buffer_state) ==
+			TX_BUFFER_STATE_RESERVED;
 }
 
-static int reserve_send_buffer_if_unused(struct icmsg_data_t *dev_data)
+static int reserve_tx_buffer_if_unused(struct icmsg_data_t *dev_data)
 {
 #ifdef CONFIG_IPC_SERVICE_ICMSG_SHMEM_ACCESS_SYNC
-	int ret = k_mutex_lock(&dev_data->send, SHMEM_ACCESS_TO);
+	int ret = k_mutex_lock(&dev_data->tx_lock, SHMEM_ACCESS_TO);
 
 	if (ret < 0) {
 		return ret;
 	}
 #endif
 
-	bool was_unused = atomic_cas(&dev_data->send_buffer_reserved,
-				  SEND_BUFFER_UNUSED, SEND_BUFFER_RESERVED);
+	bool was_unused = atomic_cas(&dev_data->tx_buffer_state,
+				  TX_BUFFER_STATE_UNUSED, TX_BUFFER_STATE_RESERVED);
 
 	return was_unused ? 0 : -EALREADY;
 }
 
-static int release_send_buffer(struct icmsg_data_t *dev_data)
+static int release_tx_buffer(struct icmsg_data_t *dev_data)
 {
-	bool was_reserved = atomic_cas(&dev_data->send_buffer_reserved,
-					SEND_BUFFER_RESERVED, SEND_BUFFER_UNUSED);
+	bool was_reserved = atomic_cas(&dev_data->tx_buffer_state,
+					TX_BUFFER_STATE_RESERVED, TX_BUFFER_STATE_UNUSED);
 
 	if (!was_reserved) {
 		return -EALREADY;
 	}
 
 #ifdef CONFIG_IPC_SERVICE_ICMSG_SHMEM_ACCESS_SYNC
-	return k_mutex_unlock(&dev_data->send);
+	return k_mutex_unlock(&dev_data->tx_lock);
 #else
 	return 0;
 #endif
 }
 
+static void release_rx_buffer(struct icmsg_data_t *dev_data, uint16_t buff_len)
+{
+	spsc_pbuf_free(dev_data->rx_ib, buff_len);
+
+#if CONFIG_IPC_SERVICE_ICMSG_NOCOPY_RX
+	dev_data->rx_buffer = NULL;
+	dev_data->rx_len = 0;
+#endif
```

Comment:
We should set the state here, shouldn't we?

``` suggestion
	dev_data->rx_len = 0;
	atomic_set(&dev_data->rx_buffer_state, RX_BUFFER_STATE_RELEASED);
#endif
}
```

Reply:
Discussed offline. It cannot be wrapped in clear way in static function